bpf/aya-bpf/src/maps/ringbuf.rs

```diff
@@ -0,0 +1,163 @@
+use core::{
+    cell::UnsafeCell,
+    mem,
+    mem::MaybeUninit,
+    ops::{Deref, DerefMut},
+};
+
+use crate::{
+    bindings::{bpf_map_def, bpf_map_type::BPF_MAP_TYPE_RINGBUF},
+    helpers::{
+        bpf_ringbuf_discard, bpf_ringbuf_output, bpf_ringbuf_query, bpf_ringbuf_reserve,
+        bpf_ringbuf_submit,
+    },
+    maps::PinningType,
+};
+
+#[repr(transparent)]
+pub struct RingBuf {
+    def: UnsafeCell<bpf_map_def>,
+}
+
+unsafe impl Sync for RingBuf {}
+
+/// A ring buffer entry, returned from [`RingBuf::reserve`].
+///
+/// You must [`submit`] or [`discard`] this entry before this gets dropped.
+///
+/// [`submit`]: RingBufEntry::submit
+/// [`discard`]: RingBufEntry::discard
+#[must_use = "BPF verifier requires ring buffer entries to be either submitted or discarded"]
+pub struct RingBufEntry<T: 'static>(&'static mut MaybeUninit<T>);
+
+impl<T> Deref for RingBufEntry<T> {
+    type Target = MaybeUninit<T>;
+
+    fn deref(&self) -> &Self::Target {
+        self.0
+    }
+}
+
+impl<T> DerefMut for RingBufEntry<T> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        self.0
+    }
+}
+
+impl<T> RingBufEntry<T> {
+    /// Discard this ring buffer entry. The entry will be skipped by the userspace reader.
+    pub fn discard(self, flags: u64) {
+        unsafe { bpf_ringbuf_discard(self.0.as_mut_ptr() as *mut _, flags) };
+    }
+
+    /// Commit this ring buffer entry. The entry will be made visible to the userspace reader.
+    pub fn submit(self, flags: u64) {
+        unsafe { bpf_ringbuf_submit(self.0.as_mut_ptr() as *mut _, flags) };
+    }
+}
+
+impl RingBuf {
+    /// Declare a BPF ring buffer.
+    ///
+    /// `max_entries` must be a power of two.
```

Comment:
Ring Buffer expect max_entries to be a power of 2 of PAGE_SIZE.

I also do believe that this shouldn't be enforce on the user face but instead be done internally like [libbpf](https://github.com/libbpf/libbpf/blob/ec6f716eda43fd0f4b865ddcebe0ce8cb56bf445/src/libbpf.c#L2463) as this break existing BPF object with a size non aligned.

Reply:
Seems mostly fine to me...

> I couldn't come up with anything more clever to keep everything const
yeah it cannot be const and need to be dynamic... I guess that's fine for now but this should really be changed to use `sysconf(_SC_PAGESIZE)` in the future.